sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldUseBaseTypes.cs

```diff
@@ -0,0 +1,333 @@
+                {
+                    paramData.ShouldReportOn = false;
+                    continue;
+                }
+
+                var symbolUsedAs = FindParameterUseAsType(identifierReference, context.SemanticModel);
+                if (symbolUsedAs != null)
+                {
+                    paramData.AddUsage(symbolUsedAs);
+                }
+            }
+
+            return parametersToCheck
+                .Values
+                .Select(p => p.GetRuleViolation())
+                .WhereNotNull();
+        }
+
+        private static bool IsTrackedParameter(IParameterSymbol parameterSymbol)
+        {
+            var type = parameterSymbol.Type;
+
+            return !type.DerivesFrom(KnownType.System_Array) &&
+                   !type.IsValueType &&
+                   !type.Is(KnownType.System_String);
+        }
+
+        private SyntaxNode GetNextUnparenthesizedParent(SyntaxNode node)
+        {
+            var expression = node as ExpressionSyntax;
+            if (expression == null)
+            {
+                return node;
+            }
+
+            var topmostParent = SyntaxHelper.GetSelfOrTopParenthesizedExpression(expression);
+            return topmostParent.Parent;
+        }
+
+        private ITypeSymbol FindParameterUseAsType(IdentifierNameSyntax identifier, SemanticModel semanticModel)
+        {
+            var identifierParent = GetNextUnparenthesizedParent(identifier);
+
+            if (identifierParent is ConditionalAccessExpressionSyntax)
+            {
+                var conditionalAccess = (ConditionalAccessExpressionSyntax)identifierParent;
+                var binding = conditionalAccess.WhenNotNull as MemberBindingExpressionSyntax;
+
+                if (binding != null)
+                {
+                    var name = binding.Name;
+                    if (name == null)
+                    {
+                        return null;
+                    }
+
+                    var accessedMember = semanticModel.GetSymbolInfo(name).Symbol;
+                    return HandlePropertyOrField(identifier, accessedMember);
+                }
+
+                var invocationExpression = conditionalAccess.WhenNotNull as InvocationExpressionSyntax;
+                if (invocationExpression != null)
+                {
+                    var memberBinding = invocationExpression.Expression as MemberBindingExpressionSyntax;
+                    if (memberBinding != null)
+                    {
+                        var invocationSymbol = semanticModel.GetSymbolInfo(memberBinding).Symbol;
+                        return HandleInvocation(identifier, invocationSymbol, semanticModel);
+                    }
+                }
+            }
+            else if (identifierParent is MemberAccessExpressionSyntax)
+            {
+                var invocationExpression = identifierParent.Parent as InvocationExpressionSyntax;
```

Comment:
Not sure there but might also be worth removing parenthesis here.
`var invocationExpression = GetNextUnparenthesizedParent(identifierParent) as InvocationExpressionSyntax`